Use fallbacks in fast primitives when eval_gpu is not implemented

zcbenz opened the pull request:
The fast primitives have fallback implementations that are used for CPU and vjp/vmap, they can also be used as placeholders for GPU backends before having a fast GPU implementation.

This PR adds a `use_fallback()` method to the fast primitives to use the fallbacks in absence of a GPU implementation.

I also cleaned up some unused properties in fast primitives.

angeloskath commented:
That is a great start! I think this should also take over the stream check and shape checks but if you want as a first step to unblock you I think it is fine to merge. Basically `use_fallback` should have something like the following signature imho

```cpp
static bool Primitive::use_fallback(
    const std::vector<array>& inputs,
    const std::vector<array>& outputs,
    const Stream& s);
```

zcbenz commented:
I think it is good idea adding parameters to `use_fallback`, though different primitives need to have different parameters.

For example `scaled_dot_product_attention` needs the mask parameters to determine whether to use fallback:
https://github.com/ml-explore/mlx/blob/55b4062dd8c71d4499a430012b49f676da91818a/mlx/fast.cpp#L742-L743

I have updated `use_fallback` to take parameters, and moved some code to `ScaledDotProductAttention::use_fallback`.

angeloskath commented on mlx/fast.cpp:
```diff
@@ -770,7 +746,9 @@ array scaled_dot_product_attention(
     inputs.push_back(broadcast_to(mask_arr, mask_shape, stream));
   }
-  if (!detail::in_grad_tracing() && implementation_supports_use_case) {
+  if (!detail::in_grad_tracing() &&
```
I think `!detail::in_grad_tracing()` should go in `use_fallback` as well. It is there because we don't have a a custom primitive for `ScaledDotProductAttentionVJP` and it is faster to use the fallback if we are computing gradients.

zcbenz replied:
That makes sense, I have updated the PR.

angeloskath commented on mlx/fast_primitives.h:
```diff
@@ -65,7 +67,6 @@ class RMSNorm : public Custom {
   }
 
  private:
-  std::function<std::vector<array>(std::vector<array>)> fallback_;
```
Thanks for that.

angeloskath reviewed (approved):
I think it looks great. I left a minor comment.
